File: src/apps/blueprint-gym/client/src/core/services/exercise-service.ts
import axios, { AxiosResponse } from "axios";
import { ExerciseState } from "../models/enums.model";
import { ExerciseFormView, ExerciseLink } from "../models/exercise.model";
import { ExerciseLookupDto } from "../models/shared.model";

const instance = axios.create({
  baseURL: "http://localhost:5001/exercise/",
});

const responseBody = (response: AxiosResponse) => response.data;

const requests = {
  get: (url: string) => instance.get(url).then(responseBody),
  post: (url: string, body: {}) => instance.post(url, body).then(responseBody),
  put: (url: string, body: {}) => instance.put(url, body).then(responseBody),
  delete: (url: string) => instance.delete(url).then(responseBody),
};

export const ExerciseService = {
  getExerciseFormView: (exerciseId: string): Promise<ExerciseFormView> => requests.get(`get/${exerciseId}`),
  saveExercise: (exercise: ExerciseFormView): Promise<ExerciseFormView> => requests.put("save", exercise),
  deleteExercise: (exerciseId: string): Promise<boolean> => requests.delete(`delete/${exerciseId}`),
  getAllExerciseLinks: (searchType: ExerciseState): Promise<ExerciseLink[]> =>
    requests.get(`search/all-links/${searchType}`),
  getAllExerciseLookups: (searchType: ExerciseState): Promise<ExerciseLookupDto[]> =>
    requests.get(`search/all-lookups/${searchType}`),
  searchExerciseLinks: (name: string, searchType: ExerciseState): Promise<ExerciseLink[]> =>
    requests.get(`search/links/${searchType}/${name}`),
  searchExerciseLookups: (name: string, searchType: ExerciseState): Promise<ExerciseLookupDto[]> =>
    requests.get(`search/lookups/${searchType}/${name}`),
};

File: src/apps/blueprint-gym/server/Business/ExerciseTracker/Services/ExerciseService.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using AutoMapper;
using BlueprintGym.Business.ExerciseTracker.Interfaces;
using BlueprintGym.Business.ExerciseTracker.Models;
using BlueprintGym.Business.Shared.Models;
using BlueprintGym.Domain.Constants.Enums;
using BlueprintGym.Domain.Core.Models;
using BlueprintGym.Domain.ExerciseTracker.Interfaces;
using BlueprintGym.Domain.ExerciseTracker.Models;
using BlueprintGym.Domain.ExerciseTracker.Options;

namespace BlueprintGym.Business.ExerciseTracker.Services
{
  public class ExerciseService : IExerciseService
  {
    private readonly IMapper mapper;
    private readonly ExerciseServiceOptions exerciseServiceOptions;
    private readonly IExerciseRepository exerciseRepository;

    public ExerciseService(
      IMapper mapper,
      ExerciseServiceOptions exerciseServiceOptions,
      IExerciseRepository exerciseRepository)
    {
      this.mapper = mapper;
      this.exerciseServiceOptions = exerciseServiceOptions;
      this.exerciseRepository = exerciseRepository;
    }

    public async Task<bool> DeleteExercise(string exerciseId)
    {
      await this.exerciseRepository.Exercise.DeleteByIdAsync(exerciseId, exerciseId).ConfigureAwait(false);
      await this.exerciseRepository.ExerciseRef.DeleteByIdAsync(exerciseId, exerciseId).ConfigureAwait(false);
      return true;
    }

    public async Task<IEnumerable<ExerciseLink>> GetAllExerciseLinks(ExerciseState searchType)
    {
      return mapper.Map<IEnumerable<ExerciseLink>>(
        (await this.exerciseRepository.ExerciseRef.GetAllAsync().ConfigureAwait(false))
          .Where(x => x.State == searchType));
    }

    public async Task<IEnumerable<ExerciseLookupDto>> GetAllExerciseLookups(ExerciseState searchType)
    {
      return mapper.Map<IEnumerable<ExerciseLookupDto>>(
        (await this.exerciseRepository.ExerciseRef.GetAllAsync().ConfigureAwait(false))
          .Where(x => x.State == searchType));
    }

    public async Task<ExerciseFormView> GetExercise(string exerciseId)
    {
      return this.mapper.Map<ExerciseFormView>(
        await this.exerciseRepository.Exercise.GetByIdAsync(exerciseId).ConfigureAwait(false));
    }

    public async Task<ExerciseFormView> SaveExercise(ExerciseFormView exercise)
    {
      var exerciseEntity = await this.exerciseRepository.Exercise.UpsertAsync(
        mapper.Map<Exercise>(exercise)).ConfigureAwait(false);

      var exerciseRefEntity = await this.exerciseRepository.ExerciseRef.UpsertAsync(
        mapper.Map<ExerciseRef>(exerciseEntity)).ConfigureAwait(false);

      return this.mapper.Map<ExerciseFormView>(exerciseEntity);
    }

    public async Task<IEnumerable<ExerciseLink>> SearchExerciseLinksByName(string name, ExerciseState searchType)
    {
      return mapper.Map<IEnumerable<ExerciseLink>>(
        (await this.exerciseRepository.ExerciseRef.GetAllAsync().ConfigureAwait(false))
          .Where(x => x.State == searchType));
    }

    public async Task<IEnumerable<ExerciseLookupDto>> SearchExerciseLookupsByName(string name, ExerciseState searchType)
    {
      return mapper.Map<IEnumerable<ExerciseLookupDto>>(
        (await this.exerciseRepository.ExerciseRef.GetAllAsync().ConfigureAwait(false))
          .Where(x => x.State == searchType));
    }
  }
}